### PR TITLE
Adds an API to return Availability Zones for pools

### PR DESCRIPTION
**Pull request by lijuncloud**
**What this PR does / why we need it**:
   this pr is added for adding an api that counting and return pools availabilityZones  without repeat

**Special notes for your reviewer**:
      

**Release note**:
  all codes had been checked  by  fakeClient

### CODE REVIEW

File: openapi-spec/swagger.yaml [hunk 1]
```diff
@@ -141,6 +141,26 @@ paths:
           description: Forbidden
         '500':
           $ref: '#/responses/HTTPStatus500'
+  '/v1beta/{projectId}/pools/availabilityZones':
```
Comment (wisererik) [inline openapi-spec/swagger.yaml]: please remove the prefix: "pools"

File: pkg/db/db.go [hunk 2]
```diff
@@ -73,6 +73,8 @@ type Client interface {
 
 	GetPool(ctx *c.Context, polID string) (*model.StoragePoolSpec, error)
 
+	GetPoolsAz(ctx *c.Context) ([]string, error)
```
Comment (wisererik) [inline pkg/db/db.go]: ->GetAvailabilityZones

File: pkg/db/drivers/etcd/etcd_test.go [hunk 3]
```diff
@@ -262,6 +266,18 @@ func TestListDocks(t *testing.T) {
 	}
 }
 
+func TestGetAvailabilityZones(t *testing.T) {
```
Comment (leonwanghui) [inline pkg/db/drivers/etcd/etcd_test.go]: It's better to be updated as `TestListAvailabilityZones`

File: pkg/db/drivers/etcd/etcd.go [hunk 4]
```diff
@@ -579,6 +579,35 @@ func (c *Client) GetPool(ctx *c.Context, polID string) (*model.StoragePoolSpec,
 	return pol, nil
 }
 
+//GetPoolsAz
+func (c *Client) GetAvailabilityZones(ctx *c.Context) ([]string, error) {
+	dbReq := &Request{
+		Url: urls.GenerateAZonesURL(urls.Etcd,""),
+	}
+	fmt.Println("GetAvailabilityZones dbReq:", dbReq)
```
Comment (leonwanghui) [inline pkg/db/drivers/etcd/etcd.go]: Do NOT print some internal message in the code, if the message is printed for debugging, please use `glog.Info()` instead.

**Comment by leonwanghui** [comment]
@lijuncloud @wisererik Is there some link for requirement description? IMO it can be considered as feature design, so please make sure it existed in our work tracking plan.

File: openapi-spec/swagger.yaml [hunk 5]
```diff
@@ -141,6 +141,26 @@ paths:
           description: Forbidden
         '500':
           $ref: '#/responses/HTTPStatus500'
+  '/v1beta/{projectId}/availabilityZones':
+    parameters:
+      - $ref: '#/parameters/projectId'
+    get:
+      tags:
+        - Pool
+      description: Lists availabilityZones without repeat for all block storage pool.
```
Comment (xing-yang) [inline openapi-spec/swagger.yaml]: s/repeat/repeating

File: pkg/api/pool.go [hunk 6]
```diff
@@ -34,6 +34,33 @@ type PoolPortal struct {
 	BasePortal
 }
 
+func (this *PoolPortal) ListAvailabilityZones() {
+	if !policy.Authorize(this.Ctx, "pool:get") {
+		return
+	}
+	azs, err := db.C.ListAvailabilityZones(c.GetContext(this.Ctx))
+	if err != nil {
+		reason := fmt.Sprintf("Get pools AvailabilityZones failed: %s", err.Error())
```
Comment (xing-yang) [inline pkg/api/pool.go]: Get AvailabilityZones for pools failed

File: pkg/api/pool_test.go [hunk 7]
```diff
@@ -66,6 +68,26 @@ var (
...
+	r, _ := http.NewRequest("GET", "/v1beta/availabilityZones", nil)
+	w := httptest.NewRecorder()
+	beego.BeeApp.Handlers.ServeHTTP(w, r)
+
+	expectedZones := "unknow"
+	t.Log(w)
+	if w.Code != 200 {
+		t.Errorf("Expected 200, actual %v", w.Code)
+	}
+//	t.Log(string(w.Body.Bytes())
```
Comment (xing-yang) [inline pkg/api/pool_test.go]: Remove this

File: pkg/db/drivers/etcd/etcd.go [hunk 8]
```diff
@@ -579,6 +579,33 @@ func (c *Client) GetPool(ctx *c.Context, polID string) (*model.StoragePoolSpec,
...
+	}
+	for _, msg := range dbRes.Message {
+		var pol = &model.StoragePoolSpec{}
+		if err := json.Unmarshal([]byte(msg), pol); err != nil {
+			log.Error("When parsing pool in db:", dbRes.Error)
+			return nil, errors.New(dbRes.Error)
+		}
+		azs = append(azs, pol.AvailabilityZone)
+	}
+	//remove repeat az
```
Comment (xing-yang) [inline pkg/db/drivers/etcd/etcd.go]: remove redundant AZ

File: pkg/db/drivers/etcd/etcd.go [hunk 9]
```diff
@@ -579,6 +579,33 @@ func (c *Client) GetPool(ctx *c.Context, polID string) (*model.StoragePoolSpec,
 	return pol, nil
 }
 
+//GetPoolsAz
```
Comment (xing-yang) [inline pkg/db/drivers/etcd/etcd.go]: ListAvailabilityZones

File: pkg/db/drivers/etcd/etcd.go [hunk 10]
```diff
@@ -579,6 +579,33 @@ func (c *Client) GetPool(ctx *c.Context, polID string) (*model.StoragePoolSpec,
...
+func (c *Client) ListAvailabilityZones(ctx *c.Context) ([]string, error) {
+	dbReq := &Request{
+		Url: urls.GeneratePoolURL(urls.Etcd, ""),
+	}
+	dbRes := c.List(dbReq)
+	if dbRes.Status != "Success" {
+		log.Error("When get pool az in db:", dbRes.Error)
```
Comment (xing-yang) [inline pkg/db/drivers/etcd/etcd.go]: Failed to get AZ for pools in db

File: pkg/utils/utils.go [hunk 11]
```diff
@@ -24,6 +24,24 @@ import (
 	log "github.com/golang/glog"
 )
 
+//remove repeat element
```
Comment (xing-yang) [inline pkg/utils/utils.go]: // remove redundant elements

File: testutils/db/fake.go [hunk 12]
```diff
@@ -99,10 +99,19 @@ func (fc *FakeDbClient) GetPool(ctx *c.Context, polID string) (*model.StoragePoo
 			return &pool, nil
 		}
 	}
-
 	return nil, errors.New("Can't find this pool resource!")
 }
 
+//GetPoolsAz
```
Comment (xing-yang) [inline testutils/db/fake.go]: ListAvailabilityZones

**Comment by xing-yang** [comment]
This should be labeled as a feature rather than a patch.

**Comment by xing-yang** [comment]
Please rewrite the release note to something like this: Adds an API to return Availability Zones for pools.

File: testutils/db/fake.go (hunk 12 above)
Comment (lijuncloud) [reply testutils/db/fake.go]: xing yang,I have changed as you said

File: openapi-spec/swagger.yaml [hunk 13]
```diff
@@ -141,6 +141,26 @@ paths:
...
+      tags:
+        - Pool
+      description: Lists availabilityZones without repeating for all block storage pool.
+      responses:
+        '200':
+          description: OK
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Pool'
```
Comment (leonwanghui) [inline openapi-spec/swagger.yaml]: From the code I saw you return a string array, but here in the api specification you specify an array of Pool struct, which is not right. Please fix this bug.

File: openapi-spec/swagger.yaml [hunk 14]
```diff
@@ -141,6 +141,25 @@ paths:
           description: Forbidden
         '500':
           $ref: '#/responses/HTTPStatus500'
+  '/v1beta/{projectId}/availabilityZones':
+    parameters:
+      - $ref: '#/parameters/projectId'
+    get:
+      tags:
+        - Pool
+      description: Lists availabilityZones without repeating for all block storage pool.
```
Comment (wisererik) [inline openapi-spec/swagger.yaml]: Lists availabilityZones without repeating for all block storage pool. -> Lists availability zones.

Comment (wisererik) [reply openapi-spec/swagger.yaml]: In addition, change the tag Pool to Availability Zone

File: pkg/api/pool.go [hunk 15]
```diff
@@ -34,6 +34,33 @@ type PoolPortal struct {
 	BasePortal
 }
 
+func (this *PoolPortal) ListAvailabilityZones() {
+	if !policy.Authorize(this.Ctx, "pool:get") {
```
Comment (wisererik) [inline pkg/api/pool.go]: "pool:get" is an api of admin role. please change it to `"availabilityZone:list"`. Then, please update the policy.json in opensds/exmaples and opensds-installer/conf: `"availabilityZone:list":""`

File: examples/policy.json [hunk 16]
```diff
@@ -45,5 +45,6 @@
   "volume_group:list": "rule:admin_or_owner",
   "volume_group:get": "rule:admin_or_owner",
   "volume_group:update": "rule:admin_or_owner",
-  "volume_group:delete": "rule:admin_or_owner"
+  "volume_group:delete": "rule:admin_or_owner",
+  "availabilityZone:list":"rule:admin_or_owner"
```
Comment (leonwanghui) [inline examples/policy.json]: It should be better if you change it to `availability_zone`.

File: examples/policy.json [hunk 17]
```diff
@@ -45,5 +45,21 @@
   "volume_group:list": "rule:admin_or_owner",
   "volume_group:get": "rule:admin_or_owner",
   "volume_group:update": "rule:admin_or_owner",
-  "volume_group:delete": "rule:admin_or_owner"
+  "volume_group:delete": "rule:admin_or_owner",
+  "availability_zone:list":""
 }
+
```
Comment (leonwanghui) [inline examples/policy.json]: Please remove these useless blank lines.

**Comment by xing-yang** [comment]
@lijuncloud Can you take a look of the CI failure?  Thanks.

**Comment by xing-yang** [comment]
Need a rebase too.

File: pkg/utils/utils.go [hunk 18]
```diff
@@ -24,6 +24,24 @@ import (
 	log "github.com/golang/glog"
 )
 
+//remove redundant elements
+func RvRepElement(arr []string) []string {
```
Comment (leonwanghui) [inline pkg/utils/utils.go]: Can you add a unit test for this function?

**Comment by leonwanghui** [comment]
Hi @xing-yang , please take a review, thx!

**Pull request merged**